design/kanister-prometheus-integration.md

```diff
@@ -0,0 +1,286 @@
+# Enhancing observability in Kanister
+
+<!-- toc -->
+- [Motivation](#motivation)
+- [Scope](#scope)
+- [Detailed Design](#detailed-design)
+  - [Architecture](#architecture)
+    - [Text description](#text-description)
+  - [Low Level APIs](#low-level-apis)
+    - [Metrics Package](#metrics-package)
+      - [Example Initialization Steps for a new CounterVec metric](#example-initialization-steps-for-a-new-countervec-metric)
+    - [Consumer Package](#consumer-package)
+- [Testing](#testing)
+<!-- /toc -->
+
+This document proposes the changes required within Kanister to integrate
+[Prometheus](https://github.com/prometheus/prometheus) and enhance the observability 
+of action sets. 
+
+## Motivation
+Kanister controller already has a registered metrics endpoint `/metrics`. There are no metrics exported other than the default Prometheus metrics that the default handler provides. Adding metrics to track the 
+ActionSets and Blueprints workflow will help improve the overall observability. 
+
+To achieve this, we need to build a framework for exporting metrics from the Kanister controller, and to 
+start with, export some metrics to Prometheus. 
+
```

Comment:
```suggestion
This framework simplifies the common need for Prometheus counters to
publish 0 values at startup for all permutations of labels and label values.
This ensures that Kanister controller restarts are recognized by Prometheus
and that the Prometheus rate() and increase() functions work properly across restarts.
 
```

Reply:
Very much agree with this comment!

Reply:
Thank you for pointing this out. I've included it now.